aiida/engine/processes/workchains/restart.py

```diff
@@ -0,0 +1,332 @@
+        self.node.set_extra('errors_handled', errors_handled)
+
+        self.report('launching {}<{}> iteration #{}'.format(self.ctx.process_name, node.pk, self.ctx.iteration))
+
+        return ToContext(children=append_(node))
+
+    def inspect_process(self):  # pylint: disable=inconsistent-return-statements,too-many-branches
+        """Analyse the results of the previous process and call the handlers when necessary.
+
+        If the process is excepted or killed, the workchain will abort. Otherwise any attached handlers will be called
+        in order of their specified priority. If the process was failed and no handler returns a report, indicating that
+        the error was handled, it is considered an unhandled process failure and the process is relaunched. If this
+        happens twice in a row, the workchain is aborted. In the case that at least one handler returned a report the
+        following matrix determines the logic that is followed:
+
+            Process  Handler    Handler     Action
+            result   report?    exit code
+            -----------------------------------------
+            Success      yes        == 0     Restart
+            Success      yes        != 0     Abort
+            Failed       yes        == 0     Restart
+            Failed       yes        != 0     Abort
+
+        If no handler returned a report and the process finished successfully, the work chain's work is considered done
+        and it will move on to the next step.
+        """
+        node = self.ctx.children[self.ctx.iteration - 1]
+
+        if node.is_excepted:
+            return self.exit_codes.ERROR_SUB_PROCESS_EXCEPTED  # pylint: disable=no-member
+
+        if node.is_killed:
+            return self.exit_codes.ERROR_SUB_PROCESS_KILLED  # pylint: disable=no-member
+
+        # Sort the handlers with a priority defined, based on their priority in reverse order
+        handlers = [handler for handler in self._handlers if handler.priority]
+        handlers = sorted(handlers, key=lambda x: x.priority, reverse=True)
+
+        last_report = None
+
+        for handler in handlers:
+
+            report = handler.method(self, node)
+
+            if report is not None and not isinstance(report, ProcessHandlerReport):
+                name = handler.method.__name__
+                raise RuntimeError('handler `{}` returned a value that is not a ProcessHandlerReport'.format(name))
+
+            # If an actual report was returned, save it so it is not overridden by next handler returning `None`
+            if report:
+                last_report = report
+
+            # After certain handlers, we may want to skip all other handlers
+            if report and report.do_break:
+                break
+
+        report_args = (self.ctx.process_name, node.pk)
+
+        # If the process failed and no handler returned a report we consider it an unhandled failure
+        if node.is_failed and not last_report:
```

Comment:
I think this condition can make problems if the user does not register a handler for some exit code. Another handler may handle a completely different problem and will provide a report for that handling. However, the actual problem wasn't resolved, so the work chain should fail.

I think one should stop